src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -146,90 +191,142 @@ private async ValueTask LoadOrReloadProjectsAsync(ImmutableSegmentedList<Project
-                        projectSystemName,
-                        loadedProjectInfo.Language,
-                        projectCreationInfo,
-                        _projectSystemHostInfo);
+                        var loadedProject = new LoadedProject(projectSystemProject, ProjectFactory.Workspace.Services.SolutionServices, _fileChangeWatcher, _targetFrameworkManager);
+                        existingProjectTargets.Add(loadedProject);
 
-                    var loadedProject = new LoadedProject(projectSystemProject, ProjectFactory.Workspace.Services.SolutionServices, _fileChangeWatcher, _targetFrameworkManager);
-                    loadedProject.NeedsReload += (_, _) => ProjectsToLoadAndReload.AddWork(projectToLoad with { ReportTelemetry = false });
-                    existingProjects.Add(loadedProject);
+                        loadedProject.NeedsReload += (_, _) => _projectsToLoadAndReload.AddWork(projectToLoad with { ReportTelemetry = false });
 
-                    (targetTelemetryInfo, targetNeedsRestore) = await loadedProject.UpdateWithNewProjectInfoAsync(loadedProjectInfo, _logger);
+                        (targetTelemetryInfo, targetNeedsRestore) = await loadedProject.UpdateWithNewProjectInfoAsync(loadedProjectInfo, hasAllInformation, _logger);
 
-                    needsRestore |= targetNeedsRestore;
-                    telemetryInfos[loadedProjectInfo] = targetTelemetryInfo with { IsSdkStyle = preferredBuildHostKind == BuildHostProcessKind.NetCore };
+                        needsRestore |= targetNeedsRestore;
+                        telemetryInfos[loadedProjectInfo] = targetTelemetryInfo with { IsSdkStyle = preferredBuildHostKind == BuildHostProcessKind.NetCore };
+                    }
                 }
-            }
 
-            if (projectToLoad.ReportTelemetry)
-            {
-                await _projectLoadTelemetryReporter.ReportProjectLoadTelemetryAsync(telemetryInfos, projectToLoad, cancellationToken);
+                foreach (var project in projectTargetsToUnload)
+                {
+                    project.Dispose();
+                    existingProjectTargets.Remove(project);
+                }
+
+                if (projectToLoad.ReportTelemetry)
+                {
+                    await _projectLoadTelemetryReporter.ReportProjectLoadTelemetryAsync(telemetryInfos, projectToLoad, cancellationToken);
+                }
+
+                if (currentLoadState is ProjectLoadState.BeginLoadingWithPrimordial(var projectId))
+                {
+                    // Remove the primordial project now that the design-time build pass is finished.
+                    await ProjectFactory.ApplyChangeToWorkspaceAsync(workspace => workspace.OnProjectRemoved(projectId), cancellationToken);
+                }
+
+                // Transition state machine
+                if (currentLoadState is ProjectLoadState.BeginLoadingWithPrimordial or ProjectLoadState.BeginLoading)
```

Comment:
This transition happens in all states -- even Loaded, as otherwise we are throwing out the update once a further reload happens.

Reply:
I will adjust this code to always assign a new state and adjust the LoadedProjectTargets list to be immutable. As-is the updates are not dropped, but, that's easy to miss because it's depending on List mutation.

Reply:
Done